packages/@lwc/engine-core/src/framework/modules/dynamic-events.ts

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (c) 2018, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+import { isUndefined, isNull, keys, propertyIsEnumerable } from '@lwc/shared';
+import { EmptyObject } from '../utils';
+import { invokeEventListener } from '../invoker';
+import type { VM } from '../vm';
+import type { VBaseElement } from '../vnodes';
+import type { RendererAPI } from '../renderer';
+
+export function patchDynamicEventListeners(
+    oldVnode: VBaseElement | null,
+    vnode: VBaseElement,
+    renderer: RendererAPI,
+    owner: VM
+) {
+    const {
+        elm,
+        data: { dynamicOn },
+    } = vnode;
+
+    // The argument passed to lwc:on is frozen.
+    // Because of this, `oldVnode.data.dynamicOn` is deeply equal
+    // to its value from the previous render cycle.
+    const oldDynamicOn =
+        isNull(oldVnode) || isUndefined(oldVnode.data.dynamicOn)
+            ? EmptyObject
+            : oldVnode.data.dynamicOn;
+    const newDynamicOn = isUndefined(dynamicOn) ? EmptyObject : dynamicOn;
+
+    if (oldDynamicOn === newDynamicOn) {
+        return;
+    }
+
+    const { addEventListener, removeEventListener } = renderer;
+    const actualEventListeners = getActualEventListeners(owner, elm!);
+
+    const oldDynamicOnNames = keys(oldDynamicOn);
+    const newDynamicOnNames = keys(newDynamicOn);
+
+    for (const name of oldDynamicOnNames) {
```

Comment:
There are a lot of very similar names flying around here, can you add comments clarifying the intent of each `for`/`if` block?